polars/polars-core/src/chunked_array/ops/rolling_window.rs

```diff
@@ -273,45 +277,125 @@ mod inner_mod {
+        };
+
+        (start, end - start)
+    }
+
     impl<T> ChunkRollApply for ChunkedArray<T>
     where
         T: PolarsNumericType,
         Self: IntoSeries,
     {
         /// Apply a rolling custom function. This is pretty slow because of dynamic dispatch.
-        fn rolling_apply(&self, window_size: usize, f: &dyn Fn(&Series) -> Series) -> Result<Self> {
-            if window_size >= self.len() {
-                return Ok(Self::full_null(self.name(), self.len()));
-            }
+        fn rolling_apply(
+            &self,
+            f: &dyn Fn(&Series) -> Series,
+            options: RollingOptions,
+        ) -> Result<Series> {
+            check_input(options.window_size, options.min_periods)?;
+
             let ca = self.rechunk();
-            let arr = ca.downcast_iter().next().unwrap();
+            if options.weights.is_some()
+                && !matches!(self.dtype(), DataType::Float64 | DataType::Float32)
+            {
+                let s = self.cast(&DataType::Float64)?;
+                return s.rolling_apply(f, options);
+            }
+
+            if options.window_size >= self.len() {
+                return Ok(Self::full_null(self.name(), self.len()).into_series());
+            }
 
+            let len = self.len();
+            let arr = ca.downcast_iter().next().unwrap();
             let series_container =
                 ChunkedArray::<T>::new_from_slice("", &[T::Native::zero()]).into_series();
             let array_ptr = &series_container.chunks()[0];
             let ptr = Arc::as_ptr(array_ptr) as *mut dyn Array as *mut PrimitiveArray<T::Native>;
             let mut builder = PrimitiveChunkedBuilder::<T>::new(self.name(), self.len());
-            for _ in 0..window_size - 1 {
-                builder.append_null();
-            }
 
-            for offset in 0..self.len() + 1 - window_size {
-                let arr_window = arr.slice(offset, window_size);
+            if let Some(weights) = options.weights {
+                let weights_series = Float64Chunked::new("weights", &weights).into_series();
+
+                let weights_series = weights_series.cast(self.dtype()).unwrap();
+
+                for idx in 0..len {
+                    let (start, size) = window_edges(idx, len, options.window_size, options.center);
+
+                    if size < options.min_periods {
+                        builder.append_null();
+                    } else {
+                        let arr_window = arr.slice(start, size);
+
+                        // Safety.
+                        // ptr is not dropped as we are in scope
+                        // We are also the only owner of the contents of the Arc
+                        // we do this to reduce heap allocs.
+                        unsafe {
+                            *ptr = arr_window;
+                        }
+
+                        let s = if size == options.window_size {
```

Comment:
This is most of the cases right?

Reply:
Yes, should I try rewrite to avoid doing this check in every iteration?